xfel/euxfel/agipd_cxigeom2nexus.py

```diff
@@ -26,176 +28,315 @@
-      for module_num in range(self.n_modules):
-        m_key = "p%d"%((quad*self.n_modules)+module_num)
-        m_name = 'AXIS_D0Q%dM%d'%(quad, module_num)
-        module_vector = self.hierarchy[q_key][m_key].local_origin.elems
-        self.create_vector(transformations, m_name, 0.0, depends_on=q_name, equipment='detector', equipment_component='detector_module',transformation_type='rotation', units='degrees', vector=(0., 0., -1.), offset = module_vector, offset_units = 'mm')
-
-        for asic_num in range(self.n_asics):
-          a_key = "p%da%d"%((quad*self.n_modules)+module_num, asic_num)
-          a_name = 'AXIS_D0Q%dM%dA%d'%(quad, module_num, asic_num)
-          asic_vector = self.hierarchy[q_key][m_key][a_key]['local_origin'].elems
-          self.create_vector(transformations, a_name, 0.0, depends_on=m_name, equipment='detector', equipment_component='detector_asic',transformation_type='rotation', units='degrees', vector=(0., 0., -1.), offset = asic_vector, offset_units = 'mm')
-
-          asicmodule = detector.create_group(array_name+'Q%dM%dA%d'%(quad,module_num,asic_num))
-          asicmodule.attrs['NX_class'] = 'NXdetector_module'
-          if self.params.mode == 'vds':
-            asicmodule.create_dataset('data_origin', (3,), data=[(quad*self.n_modules)+module_num, asic_slow*asic_num, 0], dtype='i')
-            asicmodule.create_dataset('data_size', (3,), data=[1, asic_slow, asic_fast], dtype='i')
-          elif self.params.mode == 'cxi':
-            asicmodule.create_dataset('data_origin', (2,), data=[asic_slow*((quad*self.n_modules*self.n_asics) + (module_num*self.n_asics) + asic_num), 0],
-                                                             dtype='i')
-            asicmodule.create_dataset('data_size', (2,), data=[asic_slow, asic_fast], dtype='i')
-
-          fast = self.hierarchy[q_key][m_key][a_key]['local_fast'].elems
-          slow = self.hierarchy[q_key][m_key][a_key]['local_slow'].elems
-          self.create_vector(asicmodule, 'fast_pixel_direction',pixel_size, depends_on=transformations.name+'/AXIS_D0Q%dM%dA%d'%(quad,module_num,asic_num),transformation_type='translation', units='mm', vector=fast, offset=(0. ,0., 0.))
-          self.create_vector(asicmodule, 'slow_pixel_direction',pixel_size, depends_on=transformations.name+'/AXIS_D0Q%dM%dA%d'%(quad,module_num,asic_num),transformation_type='translation', units='mm', vector=slow, offset=(0., 0., 0.))
-
-    f.close()
-
-if __name__ == '__main__':
-  import sys
-  nexus_helper = agipd_cxigeom2nexus(sys.argv[1:])
-  nexus_helper.create_nexus_master_file()
+    """
+        output_file_name = os.path.splitext(self.params.cxi_file)[0] + "_master.h5"
+        f = h5py.File(output_file_name, "w")
+        entry = f.create_group("entry")
+        entry.attrs["NX_class"] = "NXentry"
+        # --> definition
+        definition_string = f"NXmx:{get_git_revision_hash()}"
```

Comment:
Hi, I ran the code on a .cxi file, and this line adds the NeXus definitions repo revision hash to the NXmx statement.  I haven't seen that before, and it does cause dxtbx.print_header to fail.  Is the hash needed?  The definition group schema is [here](https://manual.nexusformat.org/classes/base_classes/NXentry.html#nxentry-definition-field), and it has a couple attributes that could be used, such as the version attribute.  There is an NXDL_VERSION file in the definitions folder that could be picked for this.

Reply:
This is a very good question, and I'm glad that you've asked!
There are two answers:
1. No, it's not needed directly, and indeed it causes some other `cctbx` pieces of software to fail. But it's just because of the way `cctbx` is written, there's nothing technically wrong to store there anything beyond those four letters "NXmx".
2. But on the other hand, yes, it is needed if we want to be consistent with definitions! Because as soon as we stick to the submodule and to the creation of NeXus from definitions automatically, at that particular moment we need to know the version of definitions (referring to a particular commit, i.e. hash), not only a name to the repo (like `NXmx`). Storing this information in a separate file makes the whole idea of NeXus as a self-contained format useless.

That's why I think it's the best place to store the version, but it's not in the NeXus standard yet. So I'd like to ask you, as a member of the NeXus committee, to think about adding this to the format itself :smiley:

Reply:
I rolled it back for a time being.